main/src/library/Fixpoint3/Phase/Stratifier.flix

```diff
@@ -0,0 +1,135 @@
+ * limitations under the License.
+ */
+
+mod Fixpoint3.Phase.Stratifier {
+    use Fixpoint3.Ast.Datalog.BodyPredicate.BodyAtom
+    use Fixpoint3.Ast.Datalog.Constraint.Constraint
+    use Fixpoint3.Ast.Datalog.Datalog.{Datalog, Model, Join}
+    use Fixpoint3.Ast.Datalog.{Datalog, Constraint, HeadPredicate, BodyPredicate, Polarity, Fixity}
+    use Fixpoint3.Ast.Datalog.HeadPredicate.HeadAtom
+    use Fixpoint3.Ast.Shared.PredSym
+    use Fixpoint3.Util.getOrCrash
+    use Fixpoint3.PrecedenceGraph.{MutGraph, getPseudoStrata}
+
+    ///
+    /// Array allowing lookup of the id-part of `PredSym` to name-part of `PredSym`.
+    ///
+    type alias PredIdToName[r: Region] = Array[String, r]
+
+    ///
+    /// Computes the stratification of `d`. The result is a map from `PredSym` to pseudo
+    /// strata and a their position in the pseudo-stratum. The number is meaningless,
+    /// except for allowing to seperate `PredSym` in the same pseudo-stratum by which
+    /// strata they belong to to.
+    ///
+    @Internal
+    pub def stratify(d: Datalog): Map[PredSym, (Int32, Int32)] = region rc {
+        let graph = MutGraph.empty(rc);
+        mkDepGraphInternal(graph, d);
+        let scc = getPseudoStrata(graph);
+        let numberOfPredSyms = MutGraph.getSize(graph);
+        let predSyms = Array.empty(rc, numberOfPredSyms);
+        collectPredSym(predSyms, d);
+        let result = MutMap.empty(rc);
+        scc |> List.forEachWithIndex(pseudoStrat -> list ->
+            list |> List.forEachWithIndex(strat -> set ->
+                set |> Set.forEach(x -> {
+                    let predSym = PredSym.PredSym(Array.get(x, predSyms), Int32.toInt64(x));
+                    MutMap.put(predSym, (pseudoStrat, strat), result)
+                })
+            )
+        );
+        MutMap.toMap(result)
+    }
+
+    ///
+    /// Insert edges into `graph` for the dependencies in the Datalog program `d`.
+    /// `A(x) :- B(x), A(x), not C(x).` inserts the edge `B -> A` and `C -> A` in `graph`.
+    ///
+    def mkDepGraphInternal(graph: MutGraph[r], d: Datalog): Unit \ r = match d {
+        case Datalog(_, rules) => Vector.forEach(precedenceHelper(graph), rules)
+        case Model(_) => ()
+        case Join(d1, d2) => mkDepGraphInternal(graph, d1); mkDepGraphInternal(graph, d2)
+    }
+
+    ///
+    /// Inserts edges into `graph` for the dependencies in the constraints `cnst`.
+    ///
+    /// See `mkDepGraphInternal`.
+    ///
+    def precedenceHelper(graph: MutGraph[r], cnst: Constraint): Unit \ r = match cnst {
+        case Constraint(head, body) => Vector.forEach(mkDepEdge(graph, head), body)
+    }
+
+    ///
+    /// Inserts edges into `graph` for the found dependency between `dst` and `src`. No
+    /// edge is inserted if `dst == src`.
+    ///
+    /// See `mkDepGraphInternal`.
+    ///
+    def mkDepEdge(graph: MutGraph[r], dst: HeadPredicate, src: BodyPredicate): Unit \ r = match (dst, src) {
```

Comment:
When it is a mutating method I would call it `add`/`insert` instead of `mk`. So `addDepEdge`